### PR TITLE
Bumploading

### PR DESCRIPTION
**Pull request by eeSPee**
## About The Pull Request

Allows items to be bumploaded into machines and smartfridges.

## Why It's Good For The Game

Allows you to load stuff from a conveyor belt directly into the machines.

## Changelog
:cl:
tweak: you can now load smartfridges and machines by using conveyor belts
tweak: smart fridges are opaque
del: replaced walls with floors on maps where smart fridges were on top.
/:cl:

### CODE REVIEW

File: code/datums/components/material_container.dm [hunk 1]
```diff
@@ -19,6 +19,7 @@
 	var/list/allowed_typecache
 	var/last_inserted_id
 	var/precise_insertion = FALSE
+	var/bump_loading = TRUE
```
Comment (ike709) [inline code/datums/components/material_container.dm]: This probably shouldn't be true by default.

File: code/datums/components/material_container.dm [hunk 2]
```diff
@@ -110,6 +112,25 @@
 	else if(I == active_held)
 		user.put_in_active_hand(I)
 
+/// Proc used for when player inserts materials
+/datum/component/material_container/proc/OnBumped(atom/source,obj/item/I)
+	if (bump_loading && istype(I))
+		var/list/tc = allowed_typecache
```
Comment (ike709) [inline code/datums/components/material_container.dm]: Do you really need to make a second list?

File: code/datums/components/material_container.dm [hunk 3]
```diff
@@ -110,6 +112,25 @@
...
+		if(disable_attackby)
+			return
+		if(I.item_flags & ABSTRACT ||(I.flags_1 & HOLOGRAM_1) || (I.item_flags & NO_MAT_REDEMPTION) || (tc && !is_type_in_typecache(I, tc)) || precondition || precise_insertion)
```
Comment (ike709) [inline code/datums/components/material_container.dm]: Please order these from least expensive to most expensive.

File: code/datums/components/material_container.dm [hunk 4]
```diff
@@ -110,6 +112,25 @@
 	else if(I == active_held)
 		user.put_in_active_hand(I)
 
+/// Proc used for when player inserts materials
+/datum/component/material_container/proc/OnBumped(atom/source,obj/item/I)
+	if (bump_loading && istype(I))
+		var/list/tc = allowed_typecache
+		if(disable_attackby)
```
Comment (ike709) [inline code/datums/components/material_container.dm]: Put this into the next check.

File: code/datums/components/material_container.dm [hunk 5]
```diff
@@ -110,6 +112,25 @@
...
+			return
+		if(I.item_flags & ABSTRACT ||(I.flags_1 & HOLOGRAM_1) || (I.item_flags & NO_MAT_REDEMPTION) || (tc && !is_type_in_typecache(I, tc)) || precondition || precise_insertion)
+			return
+		. = COMPONENT_NO_AFTERATTACK
+		var/material_amount = get_item_material_amount(I)
+		if(!material_amount || !has_space(material_amount))
+			return
+		var/inserted = insert_item(I)
+		if(inserted)
+			if(!istype(I, /obj/item/stack))
```
Comment (ike709) [inline code/datums/components/material_container.dm:126]: 🤔

File: code/datums/components/material_container.dm [hunk 6]
```diff
@@ -110,6 +112,25 @@
 	else if(I == active_held)
 		user.put_in_active_hand(I)
 
+/// Proc used for when player inserts materials
+/datum/component/material_container/proc/OnBumped(atom/source,obj/item/I)
```
Comment (Archanial) [inline code/datums/components/material_container.dm]:
```suggestion
/datum/component/material_container/proc/OnBump(atom/source, obj/item/I)
```
OnBump is much better name

File: code/datums/components/material_container.dm (hunk 1 above)
Comment (eeSPee) [reply code/datums/components/material_container.dm]: Yea, it allows lathe and machines that requite material input to be bumploaded.

Comment (PowerfulBacon) [reply code/datums/components/material_container.dm]: Wouldn't this allow literally any machine that uses this to be bumploaded by default, Ore silos, Protolathes, destructive analyzers etc.

Comment (eeSPee) [reply code/datums/components/material_container.dm]: Edit: Scratch that. They won't cause of another condition called precise_insertion, which makes bumploading condition kinda obsolete now that I think about it...

Comment (ike709) [reply code/datums/components/material_container.dm]: So remove it then.

**Comment by Isy232** [comment]
Factorio station, epic!

File: code/datums/components/material_container.dm (hunk 1 above)
Comment (eeSPee) [reply code/datums/components/material_container.dm]: Also there's an issue with some freezers placed on top on walls; specifically, items don't bump into them at all. Not sure how to fix this.

Comment (francinum) [reply code/datums/components/material_container.dm]: Map design issue.

Comment (eeSPee) [reply code/datums/components/material_container.dm]: Should I adress it in this PR?

Comment (eeSPee) [reply code/datums/components/material_container.dm]: IIRC this map fix was made so that you could not see through freezers the other side of the wall.

Comment (utilbird) [reply code/datums/components/material_container.dm]: You could add a bump check for closed turfs that checks for machines on the same tile

Comment (eeSPee) [reply code/datums/components/material_container.dm]: I solved this by making freezers opaque and removing the overlapping walls.

**Pull request closed without merging**